### PR TITLE
Free diskspace check: Linux, MacOS & Windows

### LINKED ISSUE #18: Do not start if there is not enough disk space available of the file system
We need at least 1Gb to safely generate and upload the recording files. 
On start, we should perform an explicit check for the available disk space in the target recording directory and fail with a meaningful message if there is not enough space.

### PR DESCRIPTION
**Pull request by neomatrix369**
Add a check for free diskspace of 1GB and then run the app if it satisfies the check

Fixes #18 

Let's discuss the environment variable settings, and other dynamics of the PR

### _Linux_
**Happy path**
```
± java -jar ./build/libs/record-and-upload-linux-0.0.20-SNAPSHOT-all.jar --config .private/aws-test-secrets --store ./build/play
INFO  [main]       - Starting recording app
INFO  [main]       - Checking diskspace
INFO  [main]       - Available disk space on the '/home/satyasai' volume (or drive): 28GB
INFO  [main]       - Start S3 Sync session
```

**Exception path**
```
± java -jar ./build/libs/record-and-upload-linux-0.0.20-SNAPSHOT-all.jar --config .private/aws-test-secrets --store ./build/play
INFO  [main]       - Starting recording app
INFO  [main]       - Checking diskspace
INFO  [main]       - Available disk space on the '/home/satyasai' volume (or drive): 28GB
ERROR [main]       - Sorry, you need at least 100GB of free disk space on this volume (or drive), in order to run the screen recording app (in either modes: video-enabled or video-disabled).
WARN  [main]       - Please make free up some disk space on this volume (or drive) and try running the screen recording app again.
```

### _Windows_
**Happy path**
```
IEUser@MSEDGEWIN10 MINGW64 /f/record-and-upload (check-free-disk-space-before-start)
$ java -jar ./build/libs/record-and-upload-windows-0.0.20-SNAPSHOT-all.jar --config .private/aws-test-secrets --store ./build/play
INFO  [main]       - Starting recording app
INFO  [main]       - Checking diskspace
INFO  [main]       - Available disk space on the 'C:\Users\IEUser' volume (or drive): 16GB
INFO  [main]       - Start S3 Sync session
```

**Exception path**
```
IEUser@MSEDGEWIN10 MINGW64 /f/record-and-upload (check-free-disk-space-before-start)
$ java -jar ./build/libs/record-and-upload-windows-0.0.20-SNAPSHOT-all.jar --config .private/aws-test-secrets --store ./build/play
INFO  [main]       - Starting recording app
INFO  [main]       - Checking diskspace
INFO  [main]       - Available disk space on the 'C:\Users\IEUser' volume (or drive): 16GB
ERROR [main]       - Sorry, you need at least 100GB of free disk space on this volume (or drive), in order to run the screen recording app (in either modes: video-enabled or video-disabled).
WARN  [main]       - Please make free up some disk space on this volume (or drive) and try running the screen recording app again.
```

The 1GB minimum requirement of disk space can be toggled to another value using the environment variable `RECORD_AND_UPLOAD_MINIMUM_DISKSPACE`. The absence of this means, it defaults to 1GB.

### CODE REVIEW

File: src/main/java/tdl/record_upload/RecordAndUploadApp.java [hunk 1]
```diff
@@ -163,6 +168,42 @@ public static void main(String[] args)  {
         }
     }
 
+    private static void checkDiskspaceRequirements() {
+        log.info("Checking diskspace");
+        int minimumRequiredDiskspaceHumanReadable = getMinimumRequiredDiskspace(); // in GB;
+        int minimumRequiredDiskspace = minimumRequiredDiskspaceHumanReadable * ONE_GB;
+        String homeDirectory = System.getProperty("user.home");
```
Comment (julianghionoiu) [inline src/main/java/tdl/record_upload/RecordAndUploadApp.java]: We are not storing the files in the home directory, but in a directory relative to the script: "./record/localstore"

File: src/main/java/tdl/record_upload/RecordAndUploadApp.java [hunk 2]
```diff
@@ -69,8 +72,10 @@
     }
 
 
-    public static void main(String[] args)  {
+    public static void main(String[] args) {
         log.info("Starting recording app");
+
+        checkDiskspaceRequirements();
```
Comment (julianghionoiu) [inline src/main/java/tdl/record_upload/RecordAndUploadApp.java]: How can we test this?

**Comment by julianghionoiu** [comment]
Hey Mani, 

Here are my comments regarding the PR:
- We should not use environment variables, the pattern we have been using is passing parameters to the Main method, we can add one for the required disk space, with a default of 1Gb.
- I like the Docket tests, let's keep them and remove the unit tests. In this context, the unit tests add more complexity with little return on investment
- The application should definitely exit with a non-zero status code if the disk space requirement is not met. A warning is not enough.

File: src/main/java/tdl/record_upload/RecordAndUploadApp.java [hunk 3]
```diff
@@ -48,6 +52,11 @@
         @Parameter(names = {"--sourcecode"}, description = "The folder that contains the source code that needs to be tracked")
         private String localSourceCodeFolder = ".";
 
+        //~~ Minimum requirements
+
+        @Parameter(names = {"--minimum-required-diskspace"}, description = "Minimum required diskspace (in GB) on the current volume (or drive) for the app to run")
```
Comment (julianghionoiu) [inline src/main/java/tdl/record_upload/RecordAndUploadApp.java]: For completeness, add "gb" to the name: "--minimum-required-diskspace-gb"

File: src/test/java/tdl/record_upload/RecordAndUploadAppTest.java [hunk 4]
```diff
@@ -52,6 +52,31 @@ void sendStop() throws Exception {
 
     @Test
     public void orchestratesMultipleThreads() throws Exception {
+        runRecordAndUploadApp();
```
Comment (julianghionoiu) [inline src/test/java/tdl/record_upload/RecordAndUploadAppTest.java]: The assertions do not make sense unless you understand what this method does. 
On that basis, it is better to have it inline. It is a fairly involved test and that needs to be read carefully.

**Review by julianghionoiu — CHANGES_REQUESTED** [review]
Looks good, only two minor comments.

File: src/main/java/tdl/record_upload/RecordAndUploadApp.java [hunk 5]
```diff
@@ -163,6 +175,35 @@ public static void main(String[] args)  {
...
+        log.info("Checking diskspace");
+        long minimumRequiredDiskspace = minimumRequiredDiskspaceHumanReadable * ONE_GB;
+        String userDirectory = System.getProperty("user.dir");
+        String userDriveOrVolume = Paths.get(userDirectory).getRoot().toString();
+        long availableDiskspace = getAvailableDiskspaceFor(userDriveOrVolume);
+        long availableDiskspaceInGB = availableDiskspace / ONE_GB;
+        float availableDiskspaceInMB = (float) (availableDiskspace / ONE_MB);
+        log.info(String.format("Available disk space on the volume (or drive) '%s': %dGB (%.3fMB)", userDriveOrVolume, availableDiskspaceInGB, availableDiskspaceInMB));
+        if (availableDiskspace < minimumRequiredDiskspace) {
+            log.error(String.format("Sorry, you need at least %dGB of free disk space on this volume (or drive), in order to run the screen recording app (in either modes: video-enabled or video-disabled).", minimumRequiredDiskspaceHumanReadable));
```
Comment (julianghionoiu) [inline src/main/java/tdl/record_upload/RecordAndUploadApp.java]: We should not mention the "video-enabled or video-disabled" modes as some of our clients require us to enforce the recording.

File: src/main/java/tdl/record_upload/RecordAndUploadApp.java (hunk 2 above)
Comment (neomatrix369) [reply src/main/java/tdl/record_upload/RecordAndUploadApp.java]: Looks like an old commit, the new ones override this change.

File: src/main/java/tdl/record_upload/RecordAndUploadApp.java (hunk 3 above)
Comment (neomatrix369) [reply src/main/java/tdl/record_upload/RecordAndUploadApp.java]: Fixed with the latest commit 63e642d

File: src/test/java/tdl/record_upload/RecordAndUploadAppTest.java (hunk 4 above)
Comment (neomatrix369) [reply src/test/java/tdl/record_upload/RecordAndUploadAppTest.java]: Fixed

File: src/main/java/tdl/record_upload/RecordAndUploadApp.java (hunk 5 above)
Comment (neomatrix369) [reply src/main/java/tdl/record_upload/RecordAndUploadApp.java]: Fixed

**Review by julianghionoiu — APPROVED** [review]

**Pull request merged**